core/trino-main/src/main/java/io/trino/json/CachingResolver.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.json;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableList;
+import io.trino.FullConnectorSession;
+import io.trino.Session;
+import io.trino.collect.cache.NonEvictableCache;
+import io.trino.json.ir.IrPathNode;
+import io.trino.metadata.BoundSignature;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.OperatorNotFoundException;
+import io.trino.metadata.ResolvedFunction;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.function.OperatorType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeManager;
+import io.trino.type.TypeCoercion;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static io.trino.json.CachingResolver.ResolvedOperatorAndCoercions.RESOLUTION_ERROR;
+import static io.trino.json.CachingResolver.ResolvedOperatorAndCoercions.operators;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A resolver providing coercions and binary operators used for JSON path evaluation,
+ * based on the operation type and the input types.
+ * <p>
+ * It is instantiated per-driver, and caches the resolved operators and coercions.
+ * Caching is applied to IrArithmeticBinary and IrComparisonPredicate path nodes.
+ * Its purpose is to avoid resolving operators and coercions on a per-row basis, assuming
+ * that the input types to the JSON path operations repeat across rows.
+ * <p>
+ * If an operator or a component coercion cannot be resolved for certain input types,
+ * it is cached as RESOLUTION_ERROR. It depends on the caller to handle this condition.
+ */
+public class CachingResolver
+{
+    private static final int MAX_CACHE_SIZE = 100;
+    private static final int MAX_ENTRIES_PER_NODE = 10;
+
+    private final Metadata metadata;
+    private final Session session;
+    private final TypeCoercion typeCoercion;
+    private final NonEvictableCache<IrPathNodeRef<IrPathNode>, NonEvictableCache<InputTypes, ResolvedOperatorAndCoercions>> operators = buildNonEvictableCache(CacheBuilder.newBuilder().maximumSize(MAX_CACHE_SIZE));
```

Comment:
Why not flatten the caches an use a composite key made of `IrPathNode ref`, `left type` and `right type`?

Reply:
Then one node could take too much space.

Reply:
That should be ok, especially if there's a case where a single node sees many combinations of types.

It would also make the code simpler. If later we find that this is not sufficient, we can think of better caching and eviction strategies, or JIT-style codegen/execution

Reply:
The cache is now flattened.